File: Skylines/ruby/skylines.rb
# Ruby 2
class Rectangle
    attr_accessor :left, :right, :height

    def initialize(specs)
        @left = specs[0].to_i
        @right = specs[2].to_i
        @height = specs[1].to_i
    end
end

if __FILE__ == $PROGRAM_NAME
    lines = STDIN.read
    lines = lines.split(/\r?\n/)
    for line in lines do
        inputs = line.chomp.split(' ')
        rectangles = Array.new
        for input in inputs do
            input = input[1..-2]
            specs = input.split(',')
            rectangles.push(Rectangle.new(specs))
        end
        height_map = Array.new((rectangles.map { |r| r.right }).max + 1, 0)
        for rect in rectangles do
            for i in (rect.left .. (rect.right - 1)).to_a do
                if height_map[i] < rect.height
                    height_map[i] = rect.height
                end
            end
        end
        moves = Array.new
        last = 0
        for i in (0..(height_map.size - 1))
            if height_map[i] != last
                moves.push(i)
                moves.push(height_map[i])
            end
            last = height_map[i]
        end
        puts (moves.map { |m| m.to_s }).join(",")
    end
end

File: Skylines/python/skylines.py
# Python 3
import sys

class Rectangle:
    def __init__(self, specs):
        self.left = int(specs[0])
        self.right = int(specs[2])
        self.height = int(specs[1])

def main():
    for line in sys.stdin:
        rectangles = []
        inputs = line.split()
        for i in inputs:
            i = i[1:-1]
            specs = i.split(",")
            rectangles.append(Rectangle(specs))
        height_map = [0] * (max(map(lambda r: r.right, rectangles)) + 1)
        for rect in rectangles:
            for i in range(rect.left, rect.right):
                if height_map[i] < rect.height:
                    height_map[i] = rect.height
        moves = []
        last = 0
        for i in range(0, len(height_map)):
            if height_map[i] != last:
                moves.append(i)
                moves.append(height_map[i])
            last = height_map[i]
        print(",".join(map(lambda x: str(x), moves)))


if __name__ == '__main__':
    main()
